shared/src/main/java/org/odk/collect/shared/injection/ObjectProvider.kt

```diff
@@ -0,0 +1,35 @@
+package org.odk.collect.shared.injection
+
+import java.util.function.Supplier
+
+/**
+ * Allows a shared singleton to act as a provider for dependencies by implementing
+ * [ObjectProviderHost] and returning a [ObjectProvider] from
+ * [ObjectProviderHost.getMultiClassProvider].
+ */
+
+interface ObjectProvider {
```

Comment:
This is a 35 line mini DI framework. I think there are definitely better solutions than this but let's finish the discussion in #4940 before thinking about that.

Reply:
I'm still thinking that if the result of #4940 is "keep Dagger" than I'll just replace this with a shared module for the Dagger setup. I'll leave this right now until we've made some decisions there.